crates/sui-core/src/quorum_driver/mod.rs

```diff
@@ -221,6 +274,163 @@ where
         self.validators.store(new_validators);
         Ok(())
     }
+
+    // TODO currently this function is not epoch-boundary-safe. We need to make it so.
+    /// Returns Ok(None) if the no conflicting transaction was retried.
+    /// Returns Ok(Some((tx_digest, true))) if one conflicting transaction was retried and succeeded,
+    /// Some((tx_digest, false)) otherwise.
+    /// Returns Error on unexpected errors.
+    #[allow(clippy::type_complexity)]
+    async fn attempt_conflicting_transactions_maybe(
+        &self,
+        good_stake: StakeUnit,
+        conflicting_tx_digests: &BTreeMap<
+            TransactionDigest,
+            (Vec<(AuthorityName, ObjectRef)>, StakeUnit),
+        >,
+        original_tx_digest: &TransactionDigest,
+    ) -> SuiResult<Option<(TransactionDigest, bool)>> {
+        let validity = self.validators.load().committee.validity_threshold();
+        // if we have >= f+1 good stake on the current transaction, no point in retrying conflicting ones
+        if good_stake >= validity {
```

Comment:
We could consider add some useful logging regarding real equivocation here.
In particular, if good_stake >= validity and the total stake at line 311 < validity, we have a proof that something equivocated. It would be useful to log that error so we could find them.

Reply:
do you mean `good_stake >= validity && total_stake >= validity` ?

Reply:
Ah right!